subsys/bluetooth/audio/stream.c

```diff
@@ -95,7 +95,48 @@ struct bt_iso_chan *bt_audio_stream_iso_chan_get(struct bt_audio_stream *stream)
 	return NULL;
 }
 
+void bt_audio_stream_cb_register(struct bt_audio_stream *stream,
```

Comment:
I know that you move code only. But I feel we need to rename this to e.g. `bt_audio_stream_init`. The reason is that there is no registration actually

Reply:
I disagree. 

This works the same way as 
```c
void bt_le_per_adv_sync_cb_register(struct bt_le_per_adv_sync_cb *cb);
void bt_conn_cb_register(struct bt_conn_cb *cb);
int bt_conn_auth_cb_register(const struct bt_conn_auth_cb *cb);
int bt_conn_auth_info_cb_register(struct bt_conn_auth_info_cb *cb);
void bt_gatt_cb_register(struct bt_gatt_cb *cb);
```
etc. (with the minor difference that many of these are registering the callbacks in a list, instead of a specific object).

Reply:
It's not the same, in the examples you mentioned there is actual callback registration, e.g. `bt_conn_cb_register` makes the `struct bt_conn_cb` to be appended to the callback lists, `bt_conn_auth_cb_register` registers `struct bt_conn_auth_cb` where can be just one registered.

In your case, the application provides uninitialized `struct bt_audio_stream *stream` object and is obliged to initialize the `ops` member by calling `bt_audio_stream_cb_register`. If the app won't do that, the object is simply unusable

Reply:
That is a fair point. 

The audio stream callbacks (or ops) are more in line with the ISO or L2CAP channel ops, which are actually never registered with a function. 

This is one the problems I've always had with both the ISO, L2CAP and `bt_audio_stream` APIs: Some fields in the structs are expected to be set directly by the upper layers, and some are expected to be controlled by the stack. This is not only not documented, but even if it was, it would be hard to keep track of. 

Ideally all the fields are the owner's to control (perhaps with exception like `sys_snode_t _node;` which are always stack controlled), or none of them are (and in which case, the struct should be opaque/abstract). 

As long as `bt_audio_stream` is this weird combination of both, it's hard to define a proper API for it. If we want `bt_audio_stream` to follow the APIs of ISO and L2CAP, then we should simply remove `bt_audio_stream_cb_register`